### ISSUE TITLE
VIT Model fails with "Statically allocated circular buffers" error

### ISSUE THREAD

**Issue by HariniMohan0102** [tenstorrent/tt-metal#10515]
**Describe the bug**
The optimized VIT high res model on WH fails with L1 buffer issue:
`E       RuntimeError: TT_THROW @ ../tt_metal/impl/program/program.cpp:489: tt::exception
E       info:
E       Statically allocated circular buffers in program 12 clash with L1 buffers on core range [(x=0,y=0) - (x=0,y=7)]. L1 buffer allocated at 1179264 and static circular buffer region ends at 1436320`

**Note:** 
- All the submodules of the optimized VIT high res model runs without error. 
- The intermediate tensors are deallocated already. 
- Checked and analysed L1 buffer report of Vit module but couldn't get conclusions from it: [VIT_operations_with_l1_buffer_report.pdf](https://github.com/user-attachments/files/16332842/VIT_operations_with_l1_buffer_report.pdf)

For further information, the PCC of submodules are here:
- patch_embeddings: 0.99
- embeddings: 0.55
- attention: 0.99
- intermediate: 0.99
- output: 0.99
- layer: 0.99
- encoder: 0.91

**To Reproduce**
1. Checkout the branch [harini/vit_on_wh](https://github.com/tenstorrent/tt-metal/compare/harini/vit_on_wh)
2. Run the following command to reproduce the issue: `pytest tests/ttnn/integration_tests/vit/test_ttnn_optimized_vit_highres_wh.py::test_vit`

**Expected behavior**
To run the optimized vit high res module without L1 buffer issue.

**Please complete the following environment information:**
- OS: [e.g. Ubuntu 20.04]
- Device: WH

**Comment by bbradelTT** [comment]
@HariniMohan0102 

tl;dr 
Please remove or comment out line 253 (`core_grid=ttnn.CoreGrid(y=8, x=8),`) in models/experimental/functional_vit/tt/ttnn_optimized_vit_highres_wh.py

Details:

There are several different optimized versions of matmul on the backend, which we refer to uOPs or matmul program configs.

If you have all inputs and the output on L1, there is going to be very little room on L1 for circular buffers (CBs).

Different versions use different sized CBs and allocate different amounts of space in L1 and have different performance characteristics.

You have 3 options:
- Do not pass in a program config and do not pass in a core grid. That makes the backend code use one approach to automatically choose a version.
- Pass in only a core grid. That makes the backend code use a different approach to automatically choose a version.
- Pass in your own program config.

Note: You can't pass in both a core grid and a program config because the program config has core grid information.

Therefore, if you run out of L1 space using one option, you can choose one of the other options.

In this case, you specified a core grid, and the option that works is specifying neither core grid nor program config. Specifying a program config is harder, and I didn't try that.

When I performed the repro steps with debug output on, I got:
```
                     Op | DEBUG    | Launching Operation: "Matmul &" (device<Tensors>)
                     Op | DEBUG    | Attributes:
                     Op | DEBUG    |    program_config = std::nullopt
                     Op | DEBUG    |    bcast_batch = 0
                     Op | DEBUG    |    output_mem_config = MemoryConfig(memory_layout=TensorMemoryLayout::INTERLEAVED,buffer_type=BufferType::L1,shard_spec=std::nullopt)
                     Op | DEBUG    |    output_dtype = DataType::BFLOAT8_B
                     Op | DEBUG    |    compute_kernel_config = WormholeComputeKernelConfig(math_fidelity=LoFi,math_approx_mode=1,fp32_dest_acc_en=0,packer_l1_acc=0) 
                     Op | DEBUG    |    untilize_out = false
                     Op | DEBUG    |    user_core_coord = (x=8,y=8)
                     Op | DEBUG    |    user_fused_activation = std::nullopt
                     Op | DEBUG    |    user_run_batched = true
                     Op | DEBUG    |    transpose_a = false
                     Op | DEBUG    |    transpose_b = false
                     Op | DEBUG    | Input Tensors:
                     Op | DEBUG    |    0: Tensor(storage=DeviceStorage(memory_config=MemoryConfig(memory_layout=TensorMemoryLayout::INTERLEAVED,buffer_type=BufferType::L1,shard_spec=std::nullopt)),shape=ttnn.Shape([1, 12, 1025[1056], 64]),dtype=DataType::BFLOAT8_B,layout=Layout::TILE)
                     Op | DEBUG    |    1: Tensor(storage=DeviceStorage(memory_config=MemoryConfig(memory_layout=TensorMemoryLayout::INTERLEAVED,buffer_type=BufferType::L1,shard_spec=std::nullopt)),shape=ttnn.Shape([1, 12, 64, 1025[1056]]),dtype=DataType::BFLOAT8_B,layout=Layout::TILE)
                     Op | DEBUG    | Optional Input Tensors:
                     Op | DEBUG    |    0: std::nullopt
                     Op | DEBUG    |
                     Op | DEBUG    | Auto generated program config: MatmulMultiCoreReuseProgramConfig(compute_with_storage_grid_size=(x=8,y=8),in0_block_w=1,out_subblock_h=1,out_subblock_w=3,per_core_M=33,per_core_N=33)
                     Op | DEBUG    | Auto generated program config: MatmulMultiCoreReuseProgramConfig(compute_with_storage_grid_size=(x=8,y=8),in0_block_w=1,out_subblock_h=1,out_subblock_w=3,per_core_M=33,per_core_N=33)
                 Always | FATAL    | Statically allocated circular buffers in program 12 clash with L1 buffers on core range [(x=0,y=0) - (x=0,y=7)]. L1 buffer allocated at 1179264 and static circular buffer region ends at 1436320



tests/ttnn/integration_tests/vit/test_ttnn_optimized_vit_highres_wh.py:462:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
models/experimental/functional_vit/tt/ttnn_optimized_vit_highres_wh.py:440: in vit
    hidden_states = vit_encoder(
models/experimental/functional_vit/tt/ttnn_optimized_vit_highres_wh.py:417: in vit_encoder
    encoder_output = vit_layer(
models/experimental/functional_vit/tt/ttnn_optimized_vit_highres_wh.py:367: in vit_layer
    multi_head_attention_output = vit_attention(
models/experimental/functional_vit/tt/ttnn_optimized_vit_highres_wh.py:248: in vit_attention
    attention_scores = ttnn.matmul(
ttnn/ttnn/decorators.py:328: in __call__
    return self.function(*function_args, **function_kwargs)
```

After commenting out line 253 (`core_grid=ttnn.CoreGrid(y=8, x=8),`) in models/experimental/functional_vit/tt/ttnn_optimized_vit_highres_wh.py, the test continued to the end and failed with a PCC error:
```
=============================================================================== short test summary info ===============================================================================
FAILED tests/ttnn/integration_tests/vit/test_ttnn_optimized_vit_highres_wh.py::test_vit[batch_size=1-sequence_size=768-image_size_w=512-image_size_h=512-model_name=google/vit-base-patch16-224] - AssertionError: 0.04298480666906725
====================================================================== 1 failed, 3 warnings in 143.94s (0:02:23) ======================================================================
```

**Comment by bbradelTT** [comment]
Looking at the output, I think the automatically chosen program config that worked was a multi cast one. The output is probably:
```
                     Op | DEBUG    | Launching Operation: "Matmul &" (device<Tensors>)
                     Op | DEBUG    | Attributes:
                     Op | DEBUG    | 	program_config = std::nullopt
                     Op | DEBUG    | 	bcast_batch = 0
                     Op | DEBUG    | 	output_mem_config = MemoryConfig(memory_layout=TensorMemoryLayout::INTERLEAVED,buffer_type=BufferType::L1,shard_spec=std::nullopt)
                     Op | DEBUG    | 	output_dtype = DataType::BFLOAT8_B
                     Op | DEBUG    | 	compute_kernel_config = WormholeComputeKernelConfig(math_fidelity=HiFi2,math_approx_mode=1,fp32_dest_acc_en=0,packer_l1_acc=0)
                     Op | DEBUG    | 	untilize_out = false
                     Op | DEBUG    | 	user_core_coord = std::nullopt
                     Op | DEBUG    | 	user_fused_activation = std::nullopt
                     Op | DEBUG    | 	user_run_batched = true
                     Op | DEBUG    | 	transpose_a = false
                     Op | DEBUG    | 	transpose_b = false
                     Op | DEBUG    | Input Tensors:
                     Op | DEBUG    | 	0: Tensor(storage=DeviceStorage(memory_config=MemoryConfig(memory_layout=TensorMemoryLayout::INTERLEAVED,buffer_type=BufferType::L1,shard_spec=std::nullopt)),shape=ttnn.Shape([1, 12, 1025[1056], 64]),dtype=DataType::BFLOAT8_B,layout=Layout::TILE)
                     Op | DEBUG    | 	1: Tensor(storage=DeviceStorage(memory_config=MemoryConfig(memory_layout=TensorMemoryLayout::INTERLEAVED,buffer_type=BufferType::L1,shard_spec=std::nullopt)),shape=ttnn.Shape([1, 12, 64, 1025[1056]]),dtype=DataType::BFLOAT8_B,layout=Layout::TILE)
                     Op | DEBUG    | Optional Input Tensors:
                     Op | DEBUG    | 	0: std::nullopt
                     Op | DEBUG    | 
                     Op | DEBUG    | Auto generated program config: MatmulMultiCoreReuseMultiCastProgramConfig(compute_with_storage_grid_size=(x=8,y=8),in0_block_w=2,out_subblock_h=1,out_subblock_w=2,per_core_M=16,per_core_N=16,transpose_mcast=0,fused_activation=std::nullopt,fuse_batch=0)
```

**Comment by HariniMohan0102** [comment]
@bbradelTT Sure, I'll remove the mentioned line and test the model. Understood about the core grid and program config. Thank you for sharing the useful information.

**Comment by bbradelTT** [comment]
> @bbradelTT Sure, I'll remove the mentioned line and test the model. Understood about the core grid and program config. Thank you for sharing the useful information.

Please let me know how it goes.
Thanks.

**Comment by bbradelTT** [comment]
@HariniMohan0102 if the error "Statically allocated circular buffers" does not occur after removing the line, please either close this issue or let me know and I'll close this issue.

**Comment by bbradelTT** [comment]
Closing. If the error "Statically allocated circular buffers" comes back even when the core_grid parameter is removed, then please re-open.

**Comment by HariniMohan0102** [comment]
"Statically allocated circular buffers" error resolved after commenting line 253 (core_grid=ttnn.CoreGrid(y=8, x=8),).
Thank you.

**Comment by bbradelTT** [comment]
As a follow up, I'm working on adding an extra check to deal with this specific corner case:
https://github.com/tenstorrent/tt-metal/pull/11892
